### PR TITLE
Apply data schema

### PR DESCRIPTION
**Pull request by mingwli**
1. apply data template when isData
2. move methods from TemplateEngine to TemplateUtility
3. keep script path in TemplateSchema class

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/4823)

### CODE REVIEW

File: src/docfx/template/TemplateEngine.cs [hunk 1]
```diff
@@ -118,7 +115,19 @@ public string GetToken(string key)
...
         public JObject TransformTocMetadata(object model)
-            => TransformMetadata("toc.json.js", JsonUtility.ToJObject(model));
+            => RunJintTransform("toc.json.js", JsonUtility.ToJObject(model));
+
+        public JObject TransformData(Document file, JObject pageModel)
+        {
+            Debug.Assert(file.IsData);
+
+            _ = _schemas.TryGetValue(file.Mime, out var templateSchema);
+            if (templateSchema.Value.HasDataTransformJs)
```
Comment (live1206) [inline src/docfx/template/TemplateEngine.cs]:
```suggestion
            if (_schemas.TryGetValue(file.Mime, out var templateSchema) && templateSchema.Value.HasDataTransformJs)
```

File: src/docfx/template/TemplateUtility.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (live1206) [inline src/docfx/template/TemplateUtility.cs]: why do we separate these methods here? they are only used in `TemplateEngine`, just because that file is too long?

File: src/docfx/template/TemplateSchema.cs [hunk 3]
```diff
@@ -28,15 +33,18 @@ public TemplateSchema(string schemaName, string schemaDir, string contentTemplat
 
             SchemaName = SchemaName;
             IsData = GetIsDataCore(schemaName, contentTemplateDir);
-            TransformMetadata = File.Exists(_metadataTransformJsPath = Path.Combine(contentTemplateDir, $"{schemaName}.mta.json.js"));
+            TransformMetadata = File.Exists(MetadataTransformJsPath = Path.Combine(contentTemplateDir, $"{schemaName}.mta.json.js"));
```
Comment (live1206) [inline src/docfx/template/TemplateSchema.cs]: `TransformMetadata` set but never used?

File: src/docfx/template/TemplateSchema.cs [hunk 4]
```diff
@@ -28,15 +33,18 @@ public TemplateSchema(string schemaName, string schemaDir, string contentTemplat
...
+
+            Debug.Assert(IsData || (!IsData && !HasDataTransformJs));
         }
 
         private bool GetIsDataCore(string schemaName, string contentTemplateDir)
         {
             if (string.Equals(schemaName, "LandingData"))
                 return false;
-            return !File.Exists(_htmlTransformJsPath = Path.Combine(contentTemplateDir, $"{schemaName}.html.primary.js"));
+            return !File.Exists(HtmlTransformJsPath = Path.Combine(contentTemplateDir, $"{schemaName}.html.primary.js"));
```
Comment (live1206) [inline src/docfx/template/TemplateSchema.cs]: where is `HtmlTransformJsPath` used?

File: src/docfx/template/TemplateSchema.cs (hunk 3 above)
Comment (live1206) [inline src/docfx/template/TemplateSchema.cs]: `MetadataTransformJsPath` set but not used?

Comment (mingwli) [reply src/docfx/template/TemplateSchema.cs]: Not yet

Comment (live1206) [reply src/docfx/template/TemplateSchema.cs]: is it going to be used in this PR or?

File: src/docfx/template/TemplateEngine.cs [hunk 5]
```diff
@@ -118,7 +115,19 @@ public string GetToken(string key)
...
 
         public JObject TransformTocMetadata(object model)
-            => TransformMetadata("toc.json.js", JsonUtility.ToJObject(model));
+            => RunJintTransform("toc.json.js", JsonUtility.ToJObject(model));
+
+        public JObject TransformData(Document file, JObject pageModel)
+        {
+            Debug.Assert(file.IsData);
+
+            _ = _schemas.TryGetValue(file.Mime, out var templateSchema);
```
Comment (OsmondJiang) [inline src/docfx/template/TemplateEngine.cs]: ? [黑人❓]

File: src/docfx/template/TemplateSchema.cs (hunk 3 above)
Comment (mingwli) [reply src/docfx/template/TemplateSchema.cs]: removed

File: src/docfx/template/TemplateEngine.cs [hunk 6]
```diff
@@ -170,11 +178,16 @@ private JObject LoadGlobalTokens(string contentTemplateDir)
             return File.Exists(path) ? JObject.Parse(File.ReadAllText(path)) : new JObject();
         }
 
-        private JObject TransformMetadata(string scriptPath, JObject model)
+        private JObject RunJintTransform(string scriptPath, JObject model)
```
Comment (OsmondJiang) [inline src/docfx/template/TemplateEngine.cs]: Too general: -> RunDataTransform
Or
Too specific -> RunJint but handle page transform.

File: src/docfx/template/TemplateEngine.cs [hunk 7]
```diff
@@ -78,11 +69,17 @@ public string Render(string content, Document file, JObject rawMetadata, string
             return _liquid.Render(layout, liquidModel);
         }
 
+        public string RenderMustache(string mime, JObject pageModel)
```
Comment (OsmondJiang) [inline src/docfx/template/TemplateEngine.cs]: Render

File: src/docfx/template/TemplateUtility.cs [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+                }
+
+                result.AppendLine($"<meta name=\"{HttpUtility.HtmlEncode(name)}\" content=\"{HttpUtility.HtmlEncode(content)}\" />");
+            }
+
+            return result.ToString();
+        }
+
+        public static JObject RemoveUpdatedAtDateTime(JObject rawMetadata)
+        {
```
[the remark this answers is not in the document]
Comment (OsmondJiang) [reply src/docfx/template/TemplateUtility.cs]: @yufeih

File: docs/specs/schema.yml [hunk 9]
```diff
@@ -319,6 +319,52 @@ outputs:
   .errors.log: |
     ["error","json-syntax-error","Invalid JavaScript property identifier character: }.","docs/a.json",1,2]
 ---
+# Run data-transform if provided, by default return {"content": "serialized json"}
```
Comment (yufeih) [inline docs/specs/schema.yml]: can you separate this change to another PR?

File: src/docfx/template/TemplateEngine.cs [hunk 10]
```diff
@@ -118,18 +115,20 @@ public string GetToken(string key)
         }
 
         public JObject TransformTocMetadata(object model)
-            => TransformMetadata("toc.json.js", JsonUtility.ToJObject(model));
+            => RunJintTransform("toc.json.js", JsonUtility.ToJObject(model));
```
Comment (mingwli) [inline src/docfx/template/TemplateEngine.cs]: remove

File: src/docfx/template/TemplateSchema.cs [hunk 11]
```diff
@@ -9,14 +9,13 @@ namespace Microsoft.Docs.Build
...
-        private string _metadataTransformJsPath;
-
         public string SchemaName { get; }
 
         public bool IsData { get; }
 
-        public bool TransformMetadata { get; }
+        public bool HasDataTransformJs { get; }
+
+        public string DataTransformJsPath { get; }
```
Comment (mingwli) [inline src/docfx/template/TemplateSchema.cs]: check in RunJint, optional

File: src/docfx/template/TemplateEngine.cs [hunk 12]
```diff
@@ -37,31 +35,28 @@ private TemplateEngine(string templateDir, JsonSchema metadataSchema)
...
-            {
-                return true;
-            }
-
-            return false;
+            return mime != null && _schemas.TryGetValue(mime, out var schemaTemplate) && schemaTemplate.Value.IsData;
         }
 
-        public TemplateSchema GetJsonSchema(string schemaName)
+        public TemplateSchema GetJsonSchema(SourceInfo<string> schemaName)
```
Comment (yufeih) [inline src/docfx/template/TemplateEngine.cs]:
```suggestion
        public TemplateSchema GetSchema(SourceInfo<string> schemaName)
```

File: src/docfx/template/TemplateEngine.cs [hunk 13]
```diff
@@ -78,21 +73,10 @@ public string Render(string content, Document file, JObject rawMetadata, string
             return _liquid.Render(layout, liquidModel);
         }
 
-        public (TemplateModel model, JObject metadata) TransformToTemplateModel(string conceptual, JObject rawMetadata, string mime)
+        public string Render(string mime, JObject pageModel)
```
Comment (yufeih) [inline src/docfx/template/TemplateEngine.cs]:
```suggestion
        public string RunMustache(string templateName, JObject pageModel)
```

File: src/docfx/template/TemplateEngine.cs [hunk 14]
```diff
@@ -117,17 +101,19 @@ public string GetToken(string key)
             return _global[key]?.ToString();
         }
 
-        public JObject TransformTocMetadata(object model)
-            => TransformMetadata("toc.json.js", JsonUtility.ToJObject(model));
-
-        public static bool IsLandingData(string mime)
+        public JObject RunJintTransform(string scriptName, JObject model)
```
Comment (yufeih) [inline src/docfx/template/TemplateEngine.cs]:
```suggestion
        public JObject RunJint(string scriptName, JObject model)
```

File: src/docfx/build/page/BuildPage.cs [hunk 15]
```diff
@@ -246,28 +243,60 @@ private static (List<Error> errors, JObject model, InputMetadata inputMetadata)
             return (errors, pageModel, inputMetadata);
         }
 
+        private static object ApplyDataTemplate(Context context, Document file, JObject pageModel)
```
Comment (live1206) [inline src/docfx/build/page/BuildPage.cs]: remove this method

File: src/docfx/build/page/BuildPage.cs [hunk 16]
```diff
@@ -246,28 +243,60 @@ private static (List<Error> errors, JObject model, InputMetadata inputMetadata)
...
                 }
 
-                return context.TemplateEngine.TransformToTemplateModel(conceptual, processedMetadata, file.Mime);
+                return TransformToTemplateModel(context, conceptual, processedMetadata, file.Mime);
             }
 
             return (pageModel, processedMetadata);
         }
+
+        private static (TemplateModel model, JObject metadata) TransformToTemplateModel(Context context, string conceptual, JObject rawMetadata, string mime)
```
Comment (live1206) [inline src/docfx/build/page/BuildPage.cs]:
```suggestion
        private static (TemplateModel model, JObject metadata) TransformToTemplateModel(Context context, string conceptual, JObject pageModel, string mime)
```

File: src/docfx/build/page/BuildPage.cs [hunk 17]
```diff
@@ -246,28 +243,60 @@ private static (List<Error> errors, JObject model, InputMetadata inputMetadata)
...
             {
-                return (context.TemplateEngine.Render(conceptual, file, processedMetadata, file.Mime), null);
+                return (context.TemplateEngine.RunLiquid(conceptual, file, processedMetadata), null);
             }
 
             if (file.Docset.Legacy)
             {
                 if (!isConceptual)
                 {
                     // run sdp JINT and mustache to generate html
```
Comment (live1206) [inline src/docfx/build/page/BuildPage.cs]: Add TODO and remove the comment out code

**Comment by mingwli** [comment]
It's caused by deleting updated_at_date_time
